Removes OO form matchrules, updates AS PATH

lukpueh opened the pull request:
- Matchrules are no longer OO
  they are simply stored as list and all the verification happens
  in verifylib
- This also fixes #14, an ugly workaround to use attr's asdict
- The commit also updates the MATCH matchrules to take an
  optional target path which allows for matching artifacts that
  have different paths

lukpueh committed:
Removes OO form matchrules, updates AS PATH

- Matchrules are no longer OO
  they are simply stored as list and all the verification happens
  in verifylib
- This also fixes #14, an ugly workaround to use attr's asdict
- The commit also updates the MATCH matchrules to take an
  optional target path which allows for matching artifacts that
  have different paths

SantiagoTorres commented on toto/verifylib.py:
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
+    (depending on the 2nd element of the rule a material or product).
+
+    Also verifies:
+    That the target_link as idetfied by ("FROM" <step>) exists in
```
identified*

SantiagoTorres commented on toto/verifylib.py:
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
+    raises an if a matchrule does not verify.
+    TBA (see https://github.com/in-toto/in-toto/issues/6)
+
+    RuleVerficationFailed
```
When this is raised would be good saying here.

lukpueh replied:
Okay, will give more info.

SantiagoTorres commented on toto/verifylib.py:
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
+  # Verify that the recorded source artifact hash and the recorded target
+  # artifact hash are equal.
+  if (ComparableHashDict(source_artifacts[source_path]) != \
+      ComparableHashDict(target_artifacts[target_path])):
```
Using these `ComparableHashDicts` is the way to go? couldn't we just compare fields within the source/target paths?

I wonder if this is an issue if, for example, one link provides two hashes and the other one provides only one.

This is a minor nit though.

lukpueh replied:
`ComparableHashDicts` is just a helper I wrote to compare two hash dicts of format `ssl_crypto.formats.HASHDICT_SCHEMA`, e.g.

``` python
{'sha256': '23432df87ab..',
'sha512': '34324abc34df..', ...}
```

The behavior is documented in its docstring:

``` python
    """Equal if the dicts have the same keys and the according values
    (strings) are equal"""
```

But we can talk about changing the behavior. Using it seems handy though.

SantiagoTorres replied:
Yeah, this was my bad. I think the current behavior is ok for the time being.

SantiagoTorres commented on toto/verifylib.py:
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
+    raises an if a matchrule does not verify.
+    TBA (see https://github.com/in-toto/in-toto/issues/6)
+
+    RuleVerficationFailed
```
same here

lukpueh replied:
Will change.

SantiagoTorres commented on toto/verifylib.py:
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
+  # FIXME: Validate rule format
+  path = rule[1]
+  if (path in link.materials.keys()):
+    raise RuleVerficationFailed("'%s' " \
```
I don't think you need the `\` here

lukpueh replied:
Will change.

SantiagoTorres commented on toto/verifylib.py:
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
+    TBA (see https://github.com/in-toto/in-toto/issues/6)
+
+  <Side Effects>
+    Calls the respective verify_<rule-type>_rule function.
```
Is this really a side-effect?

lukpueh replied:
You are right. I don't think it is a side-effect either. I guess I saw some similar "side effects" in the tuf projects and adapted the broad understanding. I'll take this out here.

SantiagoTorres commented on toto/verifylib.py:
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
 
   <Side Effects>
-    Calls matchrule verification methods.
+    Calls helper functions to verify all matchrules of an item.
```
Is this a side-effect?

lukpueh replied:
See https://github.com/in-toto/in-toto/pull/31#discussion_r86243521

SantiagoTorres reviewed (changes requested):
Some minor nits only. Thanks!

lukpueh committed:
Fixes docstrings in verifylib

- Removes wrong <Side Effect> descriptions
- Adds more info to <Exceptions>
- Fixes minor typos

lukpueh committed:
Removes unneeded linebreak escapes

SantiagoTorres commented:
👍 Thanks!
